src/mpid/common/shm/mpidu_init_shm_alloc.c

```diff
@@ -0,0 +1,652 @@
+    struct alloc_elem *next;
+    void **ptr_p;
+    size_t len;
+} alloc_elem_t;
+
+static struct {
+    alloc_elem_t *head, *tail;
+} allocq = {
+0};
+
+static int check_alloc(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t * barrier,
+                       int num_local, int local_rank);
+
+#define ALLOCQ_HEAD() GENERIC_Q_HEAD(allocq)
+#define ALLOCQ_EMPTY() GENERIC_Q_EMPTY(allocq)
+#define ALLOCQ_ENQUEUE(ep) GENERIC_Q_ENQUEUE(&allocq, ep, next)
+#define ALLOCQ_DEQUEUE(epp) GENERIC_Q_DEQUEUE(&allocq, epp, next)
+
+static size_t segment_len = 0;
+
+static int num_segments = 0;
+
+typedef struct asym_check_region {
+    void *base_ptr;
+    OPA_int_t is_asym;
+} asym_check_region;
+
+static asym_check_region *asym_check_region_p = NULL;
+
+/* MPIDU_Init_shm_seg_alloc(len, ptr_p)
+
+   This function is used to allow the caller to reserve a len sized
+   region in the shared memory segment.  Once the shared memory
+   segment is actually allocated, when MPIDU_Init_SHM_Seg_commit() is
+   called, the pointer *ptr_p will be set to point to the reserved
+   region in the shared memory segment.
+
+   Note that no shared memory is actually allocated by this function,
+   and the *ptr_p pointer will be valid only after
+   MPIDU_Init_SHM_Seg_commit() is called.
+*/
+int MPIDU_Init_shm_seg_alloc(size_t len, void **ptr_p, MPL_memory_class class)
+{
+    int mpi_errno = MPI_SUCCESS;
+    alloc_elem_t *ep;
+    MPIR_CHKPMEM_DECL(1);
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_INIT_SHM_SEG_ALLOC);
+
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDU_INIT_SHM_SEG_ALLOC);
+
+    /* round up to multiple of 8 to ensure the start of the next
+     * region is 64-bit aligned. */
+    len = MPL_ROUND_UP_ALIGN(len, (size_t) 8);
+    MPIR_Assert(len);
+    MPIR_Assert(ptr_p);
+
+    MPIR_CHKPMEM_MALLOC(ep, alloc_elem_t *, sizeof(alloc_elem_t), mpi_errno, "el", class);
+
+    ep->ptr_p = ptr_p;
+    ep->len = len;
+
+    ALLOCQ_ENQUEUE(ep);
+
+    segment_len += len;
+
+  fn_exit:
+    MPIR_CHKPMEM_COMMIT();
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDU_INIT_SHM_SEG_ALLOC);
+    return mpi_errno;
+  fn_fail:
+    MPIR_CHKPMEM_REAP();
+    goto fn_exit;
+}
+
+/* MPIDU_Init_shm_seg_commit(memory, num_local, local_rank)
+
+   This function allocates a shared memory segment large enough to
+   hold all of the regions previously requested by calls to
+   MPIDU_Init_shm_seg_alloc().  For each request, this function sets the
+   associated pointer to point to the reserved region in the allocated
+   shared memory segment.
+
+   If there is only one process local to this node, then a shared
+   memory region is not allocated.  Instead, memory is allocated from
+   the heap.
+
+   At least one call to MPIDU_Init_SHM_Seg_alloc() must be made before
+   calling MPIDU_Init_SHM_Seg_commit().
+ */
+int MPIDU_Init_shm_seg_commit(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t ** barrier,
+                              int num_local, int local_rank, int local_procs_0, int rank,
```

Comment:
I have refactored this function in PR #4036. This is fine here. We can do the rebase later whichever PR gets merged first.

Reply:
In reference to the refactoring of `MPIDU_Init_shm_seg_alloc` comment above, I think it will be easier if your PMI refactoring PR introducing `MPIR_pmi_bcast` gets merged first. Afterwards, I will remove `MPIDU_Init_shm_seg_alloc` completely and move the `core` shmem seg allocation inside `MPIDU_Init_shm_init`. I will also move `MPIDU_shm_barrier` from `mpidu_shm_barrier.c` to `mpidu_init_shm.c` and rename it to `Init_shm_barrier`.